src/Cardano/Fee.hs

```diff
@@ -254,29 +267,94 @@ reduceSingleChange (Fee fee, Coin chng)
     | otherwise =
           Coin 0
 
--- | Proportionally divide the fee over each output.
+-- | Distribute the given fee over the given list of coins, so that each coin
+--   is allocated a __fraction__ of the fee in proportion to its relative size.
 --
--- Pre-condition 1: The given outputs list shouldn't be empty
--- Pre-condition 2: None of the outputs should be null
+-- == Pre-condition
 --
--- It returns the a list of pairs (fee, output).
-divvyFee :: Fee -> [Coin] -> [(Fee, Coin)]
-divvyFee _ outs | (Coin 0) `elem` outs =
-    error "divvyFee: some outputs are null"
-divvyFee (Fee f0) outs = go f0 [] outs
+-- Every coin in the given list must be __non-zero__.
+--
+-- == Examples
+--
+-- >>> distributeFee (Fee 2) [(Coin 1), (Coin 1)]
+-- [(Fee 1, Coin 1), (Fee 1, Coin 1)]
+--
+-- >>> distributeFee (Fee 4) [(Coin 1), (Coin 1)]
+-- [(Fee 2, Coin 1), (Fee 2, Coin 1)]
+--
+-- >>> distributeFee (Fee 7) [(Coin 1), (Coin 2), (Coin 4)]
+-- [(Fee 1, Coin 1), (Fee 2, Coin 2), (Fee 4, Coin 4)]
+--
+-- >>> distributeFee (Fee 14) [(Coin 1), (Coin 2), (Coin 4)]
+-- [(Fee 2, Coin 1), (Fee 4, Coin 2), (Fee 8, Coin 4)]
+--
+distributeFee :: Fee -> NonEmpty Coin -> NonEmpty (Fee, Coin)
+distributeFee _ outs | Coin 0 `F.elem` outs =
+    error "distributeFee: one or more coins has a value of zero."
```

Comment:
maybe it is worth to use `invariant` here as we did extensively in cardano-wallet core?

Reply:
> maybe it is worth to use `invariant` here as we did extensively in cardano-wallet core?

Fixed in https://github.com/input-output-hk/cardano-coin-selection/pull/25/commits/197433953777c47278a659f44caba6f734df103e.